include/nbl/video/alloc/SubAllocatedDescriptorSet.h

```diff
@@ -0,0 +1,120 @@
+
+				for (uint32_t j = m_allocatableRanges.size(); j < binding.data; j++)
+				{
+					m_allocatableRanges.push_back({});
+				}
+
+				SubAllocDescriptorSetRange range;
+				range.reservedSize = 0;
+				// Only bindings with these flags will be allocatable
+				if (flags.hasFlags(core::bitflag(IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_UPDATE_AFTER_BIND_BIT)
+					| IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_UPDATE_UNUSED_WHILE_PENDING_BIT
+					| IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_PARTIALLY_BOUND_BIT))
+				{
+					range.reservedSize = AddressAllocator::reserved_size(maxAllocatableAlignment, static_cast<size_type>(count), args...);
+					range.reservedAllocator = std::shared_ptr<ReservedAllocator>(new ReservedAllocator());
+					range.addressAllocator = std::shared_ptr<AddressAllocator>(new AddressAllocator(
+						range.reservedAllocator->allocate(range.reservedSize, _NBL_SIMD_ALIGNMENT),
+						static_cast<size_type>(0), 0u, maxAllocatableAlignment, static_cast<size_type>(count), std::forward<Args>(args)...
+					));
+				}
+				m_allocatableRanges.insert(m_allocatableRanges.begin() + binding.data, range);
+			}
+		}
+	}
+
+	~SubAllocatedDescriptorSet()
+	{
+		for (uint32_t i = 0; i < m_allocatableRanges.size(); i++)
+		{
+			auto& range = m_allocatableRanges[i];
+			if (range.reservedSize == 0)
+				continue;
+
+			auto ptr = reinterpret_cast<const uint8_t*>(core::address_allocator_traits<AddressAllocator>::getReservedSpacePtr(*range.addressAllocator));
+			range.reservedAllocator->deallocate(const_cast<uint8_t*>(ptr), range.reservedSize);
+		}
+	}
+
+	// amount of bindings in the descriptor set layout used
+	uint32_t getLayoutBindingCount() { return m_allocatableRanges.size(); }
+
+	// whether that binding index can be sub-allocated
+	bool isBindingAllocatable(uint32_t binding) { return m_allocatableRanges[binding].reservedSize != 0; }
+
+	AddressAllocator& getBindingAllocator(uint32_t binding) 
+	{ 
+		assert(isBindingAllocatable(binding)); // Check if this binding has an allocator
+		return *m_allocatableRanges[binding].addressAllocator; 
+	}
+
+	// main methods
+
+	//! Warning `outAddresses` needs to be primed with `invalid_value` values, otherwise no allocation happens for elements not equal to `invalid_value`
+	template<typename... Args>
+	inline void multi_allocate(uint32_t binding, uint32_t count, value_type* outAddresses, const size_type* sizes, const Args&... args)
+	{
+		core::address_allocator_traits<AddressAllocator>::multi_alloc_addr(getBindingAllocator(binding), count, outAddresses, sizes, 1, args...);
+	}
+	inline void multi_deallocate(uint32_t binding, uint32_t count, const size_type* addr, const size_type* sizes)
+	{
+		auto& range = m_allocatableRanges[binding];
+		assert(range.reservedSize); // Check if this binding has an allocator
+		core::address_allocator_traits<AddressAllocator>::multi_free_addr(getBindingAllocator(binding), count, addr, sizes);
+	}
```

Comment:
you need a functor like this https://github.com/Devsh-Graphics-Programming/Nabla/blob/bd8b9dcd2e1906dff4a6863f88e0874025a33a79/include/nbl/video/alloc/CAsyncSingleBufferSubAllocator.h#L36

which in the `operator()` will wipe/write/update the descriptor bindings that have been freed, and then run the `free_addr` on the bindings

you should only have one dynamic array, which is just to store the list of bindings & indices you want to free

P.S. I recommend adding some handling of descriptor set writes with `nullptr` descriptors, because if you don't do this, the IReferenceCounted in the binding will never get dropped.

Reply:
btw your `DeferredFreeFunctor::operator()` needs to come in two overloads:
- `void` that just does the deallocation (note that the destructor DOES NOT call `operator()`)
- `bool` that does same as `void` but also takes a count of allocations we want to free up as a reference, you return true if you freed more than was there or 0, and false if there's some left (don't forget to update the counter)

The reason for this is because of the `poll` and `wait` functors of the event handler, you don't want to poll all events or wait till timeout just because some event on the queue wasn't ready. You want to bail as soon as you've made "enough space"